// sgm-backend/src/models/AMarche.js
const mongoose = require('mongoose');
const { string } = require('prop-types');

const detailsPrixSchema = new mongoose.Schema({
  reference: String,
  designation: String,
  typeProduit: String,
  marque: String,
  quantiteLivree: Number,
  dateReception: String,
  sn: String,
  finGarantie: String,
  codeBarre: String,
  memo: String,
  description: String
});

const detailProjetSchema = new mongoose.Schema({
  numeroPrix: String,
  unite: String,
  quantite: Number,
  prixUnitaire: Number,
  objetPrix: String,
  numLot: String,
  prixTotalHTVA: Number,
  detailsPrix: [detailsPrixSchema]
});

const amarcheSchema = new mongoose.Schema({
  marcheBC: String,
  date: String,
  anneeMarche: String,
  typeMarche: String,
  familleProjet: String,
  intituleProjet: String,
  fournisseur: String,
  demandeur: String,
  ficheProjet: String,
  descriptionProjet: String,
  garantie: String,
  delaiExecution: String,
  numAO: String,
  jde: String,
  detailProjet: [detailProjetSchema]
}, { timestamps: true });

module.exports = mongoose.model('AMarche', amarcheSchema);
